context.go

```diff
@@ -21,10 +25,137 @@ func AddToContext(ctx context.Context, clock Clock) context.Context {
 	return context.WithValue(ctx, keyClock, clock)
 }
 
-// FromContext extracts a clock from the context. If not present, a real clock is returned.
+// FromContext extracts a clock from the context. If not present, a real clock
+// is returned.
 func FromContext(ctx context.Context) Clock {
 	if clock, ok := ctx.Value(keyClock).(Clock); ok {
 		return clock
 	}
 	return NewRealClock()
 }
+
+// WithDeadline returns a context with a deadline based on a [FakeClock].
+//
+// The returned context ignores parent cancelation if the parent was cancelled
+// with a [context.DeadlineExceeded] error. Any other error returned by the
+// parent is treated normally, cancelling the returned context.
+//
+// If the parent is cancelled with a [context.DeadlineExceeded] error, the only
+// way to then cancel the returned context is by calling the returned
+// context.CancelFunc.
+func WithDeadline(parent context.Context, clock *FakeClock, t time.Time) (context.Context, context.CancelFunc) {
```

Comment:
Hey. I'm trying to understand how it supposed to be used.

Case that I have in mind:
```
func Handler(ctx context.Context) {
   ctx := context.WithTimeout(ctx, 5*time.Second)
   ...
   // do some method with a maximum of 5 seconds timeout.
   select {
      case <-readyChan:
         // finished work
      case ctx.Done():
         // handling timeout
   }
}
```

That is why in my PR I proposed a generic method that inputs Clock, not *FakeClock.

I wonder which usage you have in mind for these methods.

Reply:
I think this is more of a question on how to inject a FakeClock into a scope.

In your case above, I'd do:

```
type Handler struct {
  fakeClock *clockwork.FakeClock
}

func (h Handler) Handle(ctx context.Context) {
  if h.fakeClock == nil {
    ctx = context.WithTimeout(ctx, 5*time.Second)
  } else {
    ctx = clockwork.WithTimeout(ctx, h.fakeClock, 5*time.Second)
  }

  ...
  // do some method with a maximum of 5 seconds timeout.
  select {
    case <-readyChan: // finished work 
    case <-ctx.Done(): // handling timeout
  }
}
```

Reply:
This means that FakeClock goes to production, which is undesirable. Usually, I pass the Clock interface used in prod and tests. Otherwise, the code path diverges between tests and production. 
In your example, h.fakeClock != nil path is impossible to test.

Reply:
> This means that FakeClock goes to production, which is undesirable.

Production code will have a `clockwork.Clock` interface that switches between a RealClock and FakeClock, so a FakeClock is "in production" in either case, it just hides behind an interface most of the time. Here it would be `nil` in production, so I wouldn't qualify that as "in production"

> Otherwise, the code path diverges between tests and production.

This always happens to some non-zero degree when you have to inject a clock. You have to communicate to your code whether to use a RealClock or a FakeClock somehow.

I don't see a single if-else statement as terribly egregious, The newly created context uses the same code path regardless of how it is initialized.

> In your example, h.fakeClock != nil path is impossible to test.

The following would work:

```
// In test
fc := clockwork.NewFakeClock()
h := Handler{fakeClock: fc}
go func() {
  h.Handle( context.Background())
} ()

// do things with fc.

// in prod - for posterity
h := Handler{}
h.Handle(context.Background())
```

---

In general, is your contention that this PR's implementation of `clockwork.WithTimeout` and `clockwork.WithDeadline` should take a `clockwork.Clock`? If so, that is not possible to do without introducing potential race conditions or type inspection to extract the `FakeClock`.

We could change this so if  `clockwork.WithTimeout` and ` clockwork.WithDeadline` receive a nil clock, they just fall back to `context.WithTimeout` and `context.WithDeadline`. Would that help assuage your concerns?

Reply:
This does not look like a nice API to me, but that might be just me.
Even typecasting inside the method would be better.

What I usually have is:
```
Handler{clock Clock}
```
which is set to a proper one in prod and a controlled one in tests. From my experience, it is the best API and separates the library's responsibility from the developer's.

Your proposal will make all library users aware of Clock and Fakeclock, which does not look like a great design.


In my initial proposal, I used a method instead of a function. That will allow separation of implementation without type assertions.

Reply:
@3vilhamster I agree with your concern. As a user of the library, it is not acceptable to reference FakeClock in usage code. I would rather implement the timed cancellation manually, or find a different library.

Reply:
FWIW, I currently use this as a drop-in replacement for `context.WithTimeout`.

```go
func WithTimeout(ctx context.Context, timeout time.Duration) (context.Context, context.CancelFunc) {
	ctx, cancel := context.WithCancel(ctx)
	go func() {
		select {
		case <-clockwork.FromContext(ctx).After(timeout):
		case <-ctx.Done():
		}
		cancel()
	}()
	return ctx, cancel
}
```

Reply:
It is pretty clear that we have some narrowing to do on the behaviors and nuances of this request. I noticed we don't actually have an issue tracker for this, so I made #94. Please continue the conversation there.

Responding below and then closing this PR until we agree on the open questions in #94.

---

> Your proposal will make all library users aware of Clock and Fakeclock, 

This is not true, only the users of this particular feature.

I personally dislike methods that type switch now that generics are supported. I'd prefer if the function takes a concrete type to communicate that it only works under certain circumstances, the caller can execute the type switch.

Still, I could be convinced otherwise and I don't personally intend to use this feature, so if you prefer to type-switch inside the function and abstract that from the caller, that is fine.

> In my initial proposal, I used a method instead of a function.

Adding bespoke receiver functions to `clockwork.Clock` that deviate from the module-level functions in the `time` package is a non-starter for 2 reasons:

1. The deviation. The `Clock` interface specifically mimics the `time` package, and we'll maintain that convention.
2. The breakage. Adding receiver functions to public interfaces often breaks users. This is secondary to point 1 above, but worth calling out.

> it is not acceptable to reference FakeClock in usage code. 

That is an acceptable opinion to hold, but it is not universally shared by all users of this library. That is to say, if the only way we can make this work explicitly and well is to reference a FakeClock, then that is the way we'll do it.